<a href="https://colab.research.google.com/github/jcjimenezb123/ProgramacionPython/blob/master/RegresionNoLineal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Regresion exponencial

El modelo exponencial se aplica a problemas de de reacciones químicas. El modelo tiene
la forma:
$$
y = ae^{bx}
$$
Donde $a$ y $b$ son los parámetros a determinar.

El Anhídrido Nítrico se descompone homogéneamente de acuerdo a la siguiente reacción.
$$
N_2O_5\rightarrow N_2O_4 +\frac{1}{2}O_2
$$
Los datos de la reacción de $N_2O_5$ en el tiempo a una temperatura de $313K$ se muestran en la siguiente tabla:

$$
\begin{array}{|l|c|c|c|c|c|c|c|c|}
\hline
tiempo(s) &0 &500& 1000& 1500& 2000& 2500& 3000\\
\hline
N_2O_5(gr/l) &0.1000& 0.0892& 0.0776& 0.0705& 0.0603& 0.0542& 0.0471\\
\hline
\end{array}
$$

Asumiendo que la reacción es de primer orden $r = kC_A$ entonces la concentración varía
de acuerdo a:
$$
C_A = C_{A0} e^{-kt}
$$
Donde $C_{A0}$ es la concentración inicial. Encontrar los valores de $C_{A0}$ y $k$ usando una regresión exponencial.

In [1]:
import numpy as np
from scipy.optimize import curve_fit

In [2]:
x = np.array([0 ,500 ,1000 ,1500 ,2000 ,2500 ,3000])
y = np.array([0.1000 ,0.0892 ,0.0776 ,0.0705 ,0.0603 ,0.0542 ,0.0471])

In [3]:
def f(t,Ca,k):
  return Ca*np.exp(-k*t)

In [4]:
param ,_= curve_fit(f,x,y,p0 =[0.1 , -0.0001])
print ( param )

[0.10040589 0.00024895]


Por lo tanto la ecuación que de la concentración de $N_2O_5$ es:

$$
C_A = 0.10040589e^{-0.00024895t}
$$

In [5]:
xs=np.linspace (x.min(),x.max())
ys=f(xs,*param)

import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=xs,y=ys,mode='lines',name='regresion',
                  line=dict(color='royalblue',dash='dot'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion exponencial',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

#Regresion potencial

El modelo potencial tiene la forma:

$$y = ax^b$$

Donde $a$ y $b$ son los parámetros a determinar

La ecuación de Bernoulli para un flujo continuo que sale de un tanque se relaciona con el flujo Q y la altura h del tanque.

$$Q=kh^a$$

Los datos del flujo como función del nivel del líquido se muestran en la siguiente tabla:

$$
\begin{array}{|l|c|c|c|c|}
\hline
Altura h& 5& 40& 70& 100\\
\hline
Flujo Q &0.2322& 0.6450& 0.820& 0.9856\\
\hline
\end{array}
$$

Estimar los valores de $k$ y $a$.

In [6]:
x = np.array([5 ,40 ,70 ,100])
y = np.array([0.2322 ,0.6450 ,0.820 ,0.9856])

In [7]:
def Q(h,k,a):
  return k*h**a

In [8]:
param ,_= curve_fit(Q,x,y)
print ( param )

[0.10975514 0.47604171]


Por lo tanto la ecuación es:

$$Q = 0.10975514h^{0.47604171}$$

In [9]:
xs=np.linspace (x.min(),x.max())
ys=Q(xs,*param)

import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=xs,y=ys,mode='lines',name='regresion',
                  line=dict(color='royalblue',dash='dot'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion potencial',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

#Regresion reciproca o hiperbolica

El modelo Recíproco o Hiperbólico tiene la forma:

$$y=\frac{1}{a+bx}$$

Donde $a$ y $b$ son los parámetros a determinar.

La ecuación de Antoine obtiene la
presión de vapor de una sustancia a distintas temperaturas.

$$
log(P) = a+\frac{b}{c+t}
$$

Los datos de la presión de vapor como función de la temperatura se muestran en la siguiente
tabla:

$$
\begin{array}{|l|c|c|c|c|c|c|c|c|c|c|c|}
\hline
T (C) &20& 22& 25& 27& 30& 34& 40& 43& 48& 51\\
\hline
log(P) (mmHg)&1.175& 1.230& 1.312& 1.365& 1.443& 1.544& 1.690& 1.761& 1.875& 1.942\\
\hline
\end{array}
$$

Estimar los valores de $a$,$b$ y $c$ de la ecuación de Antoine.

In [10]:
x = np.array([20 ,22 ,25 ,27 ,30 ,34 ,40 ,43 ,48 ,51])
y = np.array([1.175 ,1.230 ,1.312 ,1.365 ,1.443 ,\
              1.544 ,1.690 ,1.761 ,1.875 ,1.942])

In [11]:
def P(t,a,b,c):
  return a+b/(c+t)

In [12]:
param ,_= curve_fit(P,x,y)
print ( param )

[    7.97152383 -1657.04353334   223.80983991]


Por lo tanto la ecuación de Antoine es:

$$
log(P) = 7.97152383-\frac{1657.04353334}{223.80983991+t}
$$


In [13]:
xs=np.linspace (x.min(),x.max())
ys=P(xs,*param)

import plotly.express as px
import plotly.graph_objs as go

traza1=go.Scatter(x=x,y=y,mode='markers',name='datos',
                  line=dict(color='salmon',dash='dashdot'))
traza2=go.Scatter(x=xs,y=ys,mode='lines',name='regresion',
                  line=dict(color='royalblue',dash='dot'))

data=[traza1,traza2]

layout=go.Layout(title='Regresion hiperbolica',
                 xaxis=dict(title='x'),
                 yaxis=dict(title='y'))
fig=go.Figure(data=data,layout=layout)
fig.show()

Los datos de una reaccion irreversible no isotermica se muestran en la siguiente tabla.
$$
\begin{array}{l|c|c|c|c|c|c|c}
C_A (gmol/l) & 0.05 & 0.55 & 0.65 & 0.75 & 0.9 & 1 & 1.05\\
\hline
T(K) & 395 & 410 & 380 & 388 & 400 & 373 & 405\\
\hline
Vel.Reaccion(gmol/l s) & 0.1561 & 2.8098 & 1.2224 & 1.9834 & 3.3847 & 1.3993 & 4.5913
\end{array}
$$
Se propone el siguiente modelo para la velocidad de reaccion:
$$
r=k_0e^{\frac{-\beta}{T}}C_A^n
$$
Con los datos de la tabla estimar $k_0$, $\beta$ y $n$ \\


In [14]:
import numpy as np
from scipy.optimize import curve_fit

x = np.array([[0.05,0.55,0.65,0.75,0.9,1,1.05],\
             [395,410,380,388,400,373,405]])
z = np.array([0.1561,2.8098,1.2224,1.9834,3.3847,1.3993,4.5913])

In [15]:
def f(x,k,b,n):
    return k*np.exp(-b/x[1])*np.power(x[0],n)

In [16]:
param,_=curve_fit(f,x,z)
print(param)

[1.52957208e+06 5.16886091e+03 9.87410538e-01]


Por lo tanto, la ecuacion para la velocidad de reaccion es:
$$
r=1529572.01e^{\frac{-5168.86089}{T}}C_A^{0.987410537}
$$

In [19]:
X = np.arange(0,1.05,0.01)
Y = np.arange(373, 405, 1)
X, Y = np.meshgrid(X, Y)
Z=f(np.array([X,Y]),*param)

In [21]:
import plotly.graph_objects as go

fig = go.Figure(data=[go.Scatter3d(x=x[0],y=x[1],z=z,
                                   mode='markers',
                                   marker=dict(size=5,  color=z),
                                   name='datos'),
                      go.Surface(z=Z,x=X,y=Y,name='regresion')])
fig.update_layout(scene = dict(
                    xaxis_title='Concentracion (gmol/l)',
                    yaxis_title='Temperatura (K)',
                    zaxis_title='Velocidad de reaccion'))
fig.show()